R/content.R

```diff
@@ -925,3 +975,57 @@ get_content_permissions <- function(content, add_owner = TRUE) {
   res <- content$permissions(add_owner = add_owner)
   parse_connectapi_typed(res, connectapi_ptypes$permissions)
 }
+
+#' Render a content item.
+#' 
+#' @description Submit a request to render a content item. Once submitted, the
+#' server runs an asynchronous process to render the content. This might be
+#' useful if content needs to be updated after its source data has changed,
+#' especially if this doesn't happen on a regular schedule.
+#' 
+#' Only valid for rendered content (e.g., most Quarto documents, Jupyter
+#' notebooks, R Markdown reports).
+#' 
+#' @param content The content item you wish to render.
+#' @return A [VariantTask] object that can be used to track completion of the render.
+#' @export
+content_render <- function(content) {
+  suppressWarnings({
+    validate_R6_class(content, "Content")
+    if (!content$is_rendered) {
+      stop(glue::glue("Render not supported for application mode: {content$content$app_mode}. Did you mean content_restart()?"))
+    }
+    render_task <- content$default_variant$render()
+    render_task$task_id <- render_task$id
+  
+    ContentTask$new(connect = content$get_connect(), content = content$get_content(), task = render_task)
+  })
+}
+
+#' Restart a content item.
+#' 
+#' @description Submit a request to restart a content item. Once submitted, the
+#' server performs an asynchronous request to kill all processes associated with
+#' the content item, starting new processes as needed. This might be useful if
+#' the application relies on data that is loaded at startup, or if its memory
+#' usage has grown over time.
+#' 
+#' Note that users interacting with certain types of applications may have their
+#' workflows interrupted.
+#' 
+#' Only valid for interactive content (e.g., applications, APIs).
+#' 
+#' @param content The content item you wish to restart.
+#' @export
+content_restart <- function(content) {
+  validate_R6_class(content, "Content")
+  if (!content$is_interactive) {
+    stop(glue::glue("Restart not supported for application mode: {content$content$app_mode}. Did you mean content_render()?"))
+  }
+  random_hash = token_hex(32)
+  env_var_name = glue::glue("_CONNECT_RESTART_{random_hash}")
+  # https://rlang.r-lib.org/reference/glue-operators.html#using-glue-syntax-in-packages
+  content$environment_set("{env_var_name}" := random_hash)
+  content$environment_set("{env_var_name}" := NA)
```

Comment:
Isn't this sufficient? 

```suggestion
  content$environment_set(`_CONNECT_RESTART` = random_hash)
```

Reply:
No, because that leaves an ugly-looking environment variable in the content environment pane — it'd add another one each time the content was restarted. We need to clean up after ourselves here.

Reply:
I see two changes:

- Removing the unset. We need that, because otherwise it leaves an ugly-looking environment variable in the content environment pane — it'd add another one each time the content was restarted. We need to clean up after ourselves here.
- I would be fine without adding the hash string to the environment. Here I'm just copying the `posit-sdk-py` implementation. @tdstein what do you think about removing the hash from the var name?

I think I see what you mean — if we just have the same variable name, maybe we don't need to clean up after ourselves, we can just update the variable each time.

I don't think it causes any _harm_ to submit two variable requests — it happens immediately and doesn't cause _extra_ process spins that we'd need to, like, debounce (unless a user manage to refresh their page in between the two requests). Either way, I prefer cleaning up after ourselves. Would be fine removing the hash from the var name (I think I like the cleaner look of that in the dashboard in case orphaned variables do occur. Don't think `_CONNECT_RESTART` is likely to collide with anything.)

Reply:
I think it will be confusing for users to see the `_CONNECT_RESTART` variable in their envvars list. 

The set/unset routine happens quickly, so it's unlikely a user would see the variable. And as @toph-allen mentioned, Connect denounces the restart side-effect within a window. So, in the rare case that set and unset happen in different windows, it's still very unlikely that a user would be able to load their application successfully before the second window completed.

The `random_hash` is an extra measure to avoid collisions with users, although it's very unlikely a user would have an envvar named  `_CONNECT_RESTART`.

Reply:
However, as I revisit this, I think using a UNIX timestamp would be better than a random hash. In that case, if something did go wrong during the `unset` routine, it's possible to track when the restart was initiated.

Reply:
Agreed on `_CONNECT_RESTART_{unix_time_stamp}` :)